File: JS/src/day-mode.js
/**
 * Hermes Dream OS - Day Mode (JavaScript)
 * 活跃期处理：Morning Briefing / Midday Check-in / Evening Reflection / Memory Consolidation
 * 融合 Hermes Life OS 的 show, don't remind 原则
 */

import { MemoryHub } from './memory-hub.js';

export const DAY_ACTIONS = {
  BRIEFING: 'briefing',
  CHECKIN: 'checkin',
  REFLECTION: 'reflection',
  CONSOLIDATE: 'consolidate',
  WIN: 'win',
  STRUGGLE: 'struggle'
};

// ========== Day Mode 主函数 ==========

/**
 * Day Mode 入口
 * @param {string} workspaceDir - 工作目录
 * @param {'briefing'|'checkin'|'reflection'|'consolidate'} action - 操作类型
 * @param {...any} args - 额外参数
 */
export function dayMode(workspaceDir, action, ...args) {
  const hub = new MemoryHub(workspaceDir);

  switch (action) {
    case DAY_ACTIONS.BRIEFING:
      return morningBriefing(hub);
    case DAY_ACTIONS.CHECKIN:
      return middayCheckin(hub, ...args);
    case DAY_ACTIONS.REFLECTION:
      return eveningReflection(hub);
    case DAY_ACTIONS.CONSOLIDATE:
      return memoryConsolidation(hub);
    case DAY_ACTIONS.WIN:
      return recordWin(hub, ...args);
    case DAY_ACTIONS.STRUGGLE:
      return recordStruggle(hub, ...args);
    default:
      throw new Error(`Unknown action: ${action}`);
  }
}

// ========== Morning Briefing (07:00) ==========

/**
 * 晨间简报
 * 基于 Hermes Life OS 的 show, don't remind 原则
 */
function morningBriefing(hub) {
  const now = new Date();
  const today = now.toISOString().split('T')[0];
  const weekday = now.toLocaleDateString('en-US', { weekday: 'long' });

  // 1. 能量预测（基于历史数据）
  const recentEnergy = hub.getRecentEnergy(14);
  const weekdayEnergy = recentEnergy.filter(e => {
    try {
      const date = new Date(e.date);
      return date.toLocaleDateString('en-US', { weekday: 'long' }) === weekday;
    } catch {
      return false;
    }
  });

  let energyLevel = '未知（数据不足）';
  if (weekdayEnergy.length > 0) {
    const avgProductivity = weekdayEnergy.reduce((sum, e) => sum + (e.productivity || 5), 0) / weekdayEnergy.length;
    if (avgProductivity >= 7) energyLevel = '高效';
    else if (avgProductivity >= 5) energyLevel = '平稳';
    else energyLevel = '需要调整';
  }

  // 2. 今日目标
  const goals = hub.readGoals();
  const activeGoals = Object.entries(goals)
    .filter(([, goal]) => (goal.progress || 0) < 1.0)
    .map(([name, goal]) => ({
      name,
      progress: goal.progress || 0,
      deadline: goal.deadline || '无'
    }))
    .sort((a, b) => a.progress - b.progress)
    .slice(0, 3);

  // 3. 一个洞察
  const insights = hub.getRecentInsights(14, 0.7);
  const topInsight = insights[0] || null;

  // 4. 今日习惯提醒（未完成的习惯）
  const habits = hub.readHabits();
  const habitReminders = Object.entries(habits)
    .filter(([, data]) => data.last_done !== today)
    .map(([name]) => name)
    .slice(0, 3);

  // 5. 最近的成就（本周）
  const wins = hub.readJsonl(hub.winsPath);
  const recentWins = wins
    .filter(w => {
      try {
        const date = new Date(w.date);
        const daysAgo = Math.floor((Date.now() - date.getTime()) / (1000 * 60 * 60 * 24));
        return daysAgo <= 7;
      } catch {
        return false;
      }
    })
    .slice(-3);

  const result = {
    type: 'morning_briefing',
    timestamp: now.toISOString(),
    sections: {
      energy_forecast: {
        weekday,
        predicted: energyLevel,
        sample_size: weekdayEnergy.length
      },
      today_goals: {
        active: activeGoals,
        total: Object.keys(goals).length
      },
      insight: topInsight
        ? {
            text: topInsight.observation,
            confidence: topInsight.confidence,
            date: topInsight.date
          }
        : { text: '暂无高置信度洞察', confidence: 0 },
      habit_reminders: habitReminders,
      recent_wins: recentWins.map(w => w.description)
    }
  };

  hub.appendEvent('morning_briefing', { date: today });

  return result;
}

// ========== Midday Check-in (12:00) ==========

/**
 * 午间检查
 * @param {MemoryHub} hub
 * @param {string|null} level - 精力水平: low, medium, high
 * @param {number|null} productivity -生产力评分 1-10
 * @param {string} note - 备注
 */
function middayCheckin(hub, level = null, productivity = null, note = '') {
  const now = new Date();

  // 如果没有提供参数，返回检查表单
  if (!level) {
    return {
      type: 'midday_checkin',
      timestamp: now.toISOString(),
      prompt: '请记录午间精力状态',
      usage: 'day-mode.js <workspace> checkin <level> [productivity] [note]',
      level_options: ['low', 'medium', 'high'],
      example: 'day-mode.js /path/to/workspace checkin high 8'
    };
  }

  const record = hub.recordEnergy(level, note, productivity);

  // 分析上午的进展
  const today = now.toISOString().split('T')[0];
  const wins = hub.readJsonl(hub.winsPath).filter(w => w.date === today);
  const struggles = hub.readJsonl(hub.strugglesPath).filter(w => w.date === today);

  hub.appendEvent('midday_checkin', { level, productivity, note });

  return {
    type: 'midday_checkin',
    timestamp: now.toISOString(),
    record,
    morning_summary: {
      wins_today: wins.length,
      struggles_today: struggles.length
    }
  };
}

// ========== Evening Reflection (18:00) ==========

/**
 * 晚间反思
 */
function eveningReflection(hub) {
  const now = new Date();
  const today = now.toISOString().split('T')[0];

  // 读取今日数据
  const wins = hub.readJsonl(hub.winsPath).filter(w => w.date === today);
  const struggles = hub.readJsonl(hub.strugglesPath).filter(w => w.date === today);
  const moods = hub.getRecentMoods(1);
  const energy = hub.getRecentEnergy(1);

  // 获取未解决的困境
  const unresolvedStruggles = struggles.filter(s => !s.resolved);

  // 计算今日成就总数
  const totalWins = wins.length;

  // 获取活跃目标进度
  const goals = hub.readGoals();
  const activeGoals = Object.entries(goals)
    .filter(([, g]) => (g.progress || 0) < 1.0)
    .map(([name, g]) => ({
      name,
      progress: g.progress || 0,
      deadline: g.deadline || '无'
    }));

  // 获取习惯连续
  const habits = hub.readHabits();
  const habitStreaks = Object.entries(habits)
    .map(([name, data]) => ({
      name,
      streak: data.streak || 0,
      completed_today: data.last_done === today
    }))
    .sort((a, b) => b.streak - a.streak)
    .slice(0, 5);

  const result = {
    type: 'evening_reflection',
    timestamp: now.toISOString(),
    sections: {
      wins_today: wins.map(w => w.description),
      struggles_today: struggles.map(s => s.description),
      unresolved_struggles: unresolvedStruggles.map(s => s.description),
      mood: moods[moods.length - 1] || null,
      energy: energy[energy.length - 1] || null,
      active_goals: activeGoals,
      habit_streaks: habitStreaks
    },
    prompt_suggestions: unresolvedStruggles.length > 0
      ? [`解决困境: ${unresolvedStruggles[0].description}`]
      : []
  };

  hub.appendEvent('evening_reflection', {
    wins_count: wins.length,
    struggles_count: struggles.length
  });

  return result;
}

// ========== Memory Consolidation (23:00) ==========

/**
 * 记忆整合
 * 存储今日模式 + 更新习惯连续
 */
function memoryConsolidation(hub) {
  const now = new Date();
  const today = now.toISOString().split('T')[0];

  // 统计今日数据
  const wins = hub.readJsonl(hub.winsPath).filter(w => w.date === today);
  const struggles = hub.readJsonl(hub.strugglesPath).filter(w => w.date === today);
  const moods = hub.getRecentMoods(1);
  const energy = hub.getRecentEnergy(1);

  // 更新习惯连续
  const habits = hub.readHabits();
  const habitUpdates = [];

  for (const [name, data] of Object.entries(habits)) {
    const completed = data.history?.some(h => h.date === today && h.completed);
    if (completed) {
      habitUpdates.push(hub.updateHabit(name, true));
    }
  }

  // 生成今日摘要
  const summary = {
    date: today,
    wins_count: wins.length,
    struggles_count: struggles.length,
    mood_avg: moods.length > 0
      ? (moods.reduce((sum, m) => sum + (m.score || 5), 0) / moods.length).toFixed(1)
      : null,
    energy_distribution: {
      high: energy.filter(e => e.level === 'high').length,
      medium: energy.filter(e => e.level === 'medium').length,
      low: energy.filter(e => e.level === 'low').length
    },
    habits_updated: Object.keys(habits).length,
    active_streaks: Object.values(habits).filter(h => h.streak > 0).length
  };

  hub.appendEvent('memory_consolidation', summary);

  return {
    type: 'memory_consolidation',
    timestamp: now.toISOString(),
    summary
  };
}

// ========== Record Win/Struggle ==========

/**
 * 记录成就
 */
function recordWin(hub, description) {
  if (!description) {
    return { error: '描述不能为空', usage: 'day-mode.js <workspace> win <description>' };
  }

  const record = hub.recordWin(description);
  hub.appendEvent('win_recorded', { description });

  return {
    success: true,
    type: 'win_recorded',
    record
  };
}

/**
 * 记录困境
 */
function recordStruggle(hub, description, resolved = false) {
  if (!description) {
    return { error: '描述不能为空', usage: 'day-mode.js <workspace> struggle <description> [resolved]' };
  }

  const record = hub.recordStruggle(description, resolved);
  hub.appendEvent('struggle_recorded', { description, resolved });

  return {
    success: true,
    type: 'struggle_recorded',
    record
  };
}

// ========== CLI 入口 ==========

// 使用方式: node day-mode.js <workspace_dir> <action> [args...]
if (import.meta.url === `file://${process.argv[1]}`) {
  const [, , workspaceDir, action, ...args] = process.argv;

  if (!workspaceDir || !action) {
    console.log(`
Hermes Dream OS - Day Mode
==========================

Usage: node day-mode.js <workspace_dir> <action> [args...]

Actions:
  briefing                    - Morning briefing
  checkin <level> [prod] [n]  - Midday check-in (level: low|medium|high)
  reflection                  - Evening reflection
  consolidate                 - Memory consolidation
  win <description>           - Record a win
  struggle <description> [r] - Record a struggle (r: true|false)

Examples:
  node day-mode.js /path/to/workspace briefing
  node day-mode.js /path/to/workspace checkin high 8
  node day-mode.js /path/to/workspace win "完成项目X"
  node day-mode.js /path/to/workspace struggle "会议太多"
`);
    process.exit(1);
  }

  try {
    const result = dayMode(workspaceDir, action, ...args);
    console.log(JSON.stringify(result, null, 2));
  } catch (e) {
    console.error('Error:', e.message);
    process.exit(1);
  }
}

File: SCRIPTS/day-mode.py
"""
Hermes Dream OS — Day Mode (Life OS)
活跃期处理：Morning Briefing / Midday Check-in / Evening Reflection / Memory Consolidation
"""

import sys
import os
from datetime import datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from memory_hub import MemoryHub


# ========== Morning Briefing ==========

def morning_briefing(workspace_dir: str) -> dict:
    """
    07:00 - 晨间简报
    基于 Hermes Life OS 的 show, don't remind 原则

    输出：
    - 能量预测
    - 今日优先级
    - 一个洞察
    """
    hub = MemoryHub(workspace_dir)
    now = datetime.now()
    today = now.strftime("%Y-%m-%d")
    weekday = now.strftime("%A")

    result = {
        "type": "morning_briefing",
        "timestamp": now.isoformat(),
        "sections": {}
    }

    # 1. 能量预测（基于历史数据）
    recent_energy = hub.get_recent_energy(days=14)
    weekday_energy = [e for e in recent_energy if e.get("date", "") and datetime.strptime(e["date"], "%Y-%m-%d").strftime("%A") == weekday]

    if weekday_energy:
        avg_productivity = sum(e.get("productivity", 5) for e in weekday_energy) / len(weekday_energy)
        if avg_productivity >= 7:
            energy_level = "高效"
        elif avg_productivity >= 5:
            energy_level = "平稳"
        else:
            energy_level = "需要调整"
    else:
        energy_level = "未知（数据不足）"

    result["sections"]["energy_forecast"] = {
        "weekday": weekday,
        "predicted": energy_level,
        "sample_size": len(weekday_energy)
    }

    # 2. 今日目标
    goals = hub.read_goals()
    active_goals = []
    for name, goal in goals.items():
        if goal.get("progress", 0) < 1.0:
            active_goals.append({
                "name": name,
                "progress": goal.get("progress", 0),
                "deadline": goal.get("deadline", "无")
            })

    # 按进度排序，最需要的先
    active_goals.sort(key=lambda x: x["progress"])

    result["sections"]["today_goals"] = {
        "active": active_goals[:3],
        "total": len(active_goals)
    }

    # 3. 一个洞察
    insights = hub.get_recent_insights(days=14, min_confidence=0.7)
    top_insight = insights[0] if insights else None

    if top_insight:
        result["sections"]["insight"] = {
            "text": top_insight["observation"],
            "confidence": top_insight["confidence"],
            "date": top_insight.get("date", "")
        }
    else:
        result["sections"]["insight"] = {
            "text": "暂无高置信度洞察",
            "confidence": 0
        }

    # 4. 今日习惯提醒
    habits = hub.read_habits()
    habit_reminders = []
    for name, data in habits.items():
        if data.get("last_done") != today:
            habit_reminders.append({
                "name": name,
                "streak": data.get("streak", 0),
                "last_done": data.get("last_done", "从未")
            })

    result["sections"]["habit_reminders"] = habit_reminders[:3]

    # 5. 事件记录
    hub.append_event(
        "life.briefing.morning",
        weekday=weekday,
        energy_level=energy_level,
        goals_count=len(active_goals)
    )

    return result


def format_briefing_text(result: dict, user_name: str = "主人") -> str:
    """
    将简报结果格式化为文本（Hermes 风格）
    """
    now = datetime.now()
    date_str = now.strftime("%Y-%m-%d")
    weekday = now.strftime("%A")

    sections = result.get("sections", {})

    lines = [
        f"早上好，{user_name}。{date_str}，{weekday}。",
        "",
    ]

    # 能量预测
    energy = sections.get("energy_forecast", {})
    predicted = energy.get("predicted", "未知")
    lines.append(f"**能量预测**")
    lines.append(f"{weekday} 对你来说通常是 {predicted} 的一天。")
    lines.append("")

    # 洞察
    insight = sections.get("insight", {})
    if insight.get("confidence", 0) > 0:
        lines.append("**一个洞察**")
        lines.append(f"> {insight['text']}")
        lines.append("")

    # 今日目标
    goals = sections.get("today_goals", {})
    active = goals.get("active", [])
    if active:
        lines.append("**今日目标**")
        for g in active[:3]:
            progress = g.get("progress", 0) * 100
            lines.append(f"-> {g['name']} ({progress:.0f}%)")
        lines.append("")

    # 习惯提醒
    reminders = sections.get("habit_reminders", [])
    if reminders:
        lines.append("**习惯提醒**")
        for h in reminders:
            lines.append(f"- {h['name']} (已连续 {h['streak']} 天)")
        lines.append("")

    return "\n".join(lines)


# ========== Midday Check-in ==========

def midday_checkin(workspace_dir: str, energy_level: str = None, productivity: int = None, note: str = "") -> dict:
    """
    12:00 - 午间检查

    参数：
    - energy_level: low | medium | high
    - productivity: 1-10
    - note: 可选备注
    """
    hub = MemoryHub(workspace_dir)
    now = datetime.now()

    result = {
        "type": "midday_checkin",
        "timestamp": now.isoformat(),
        "recorded": False
    }

    if energy_level or productivity:
        if energy_level:
            hub.record_energy(level=energy_level, context=note, productivity=productivity)
            result["recorded"] = True
            result["energy_level"] = energy_level

        if productivity:
            result["productivity"] = productivity

        hub.append_event(
            "life.checkin.midday",
            energy_level=energy_level,
            productivity=productivity,
            note=note
        )

    return result


# ========== Evening Reflection ==========

def evening_reflection(workspace_dir: str) -> dict:
    """
    18:00 - 晚间反思

    展示：完成事项 / 未完成 / 模式观察
    """
    hub = MemoryHub(workspace_dir)
    now = datetime.now()
    today = now.strftime("%Y-%m-%d")

    result = {
        "type": "evening_reflection",
        "timestamp": now.isoformat(),
        "sections": {}
    }

    # 1. 今日成就
    wins = hub.read_jsonl(hub.wins_path)
    today_wins = [w for w in wins if w.get("date") == today]
    result["sections"]["wins"] = today_wins

    # 2. 今日困境
    struggles = hub.read_jsonl(hub.struggles_path)
    today_struggles = [s for s in struggles if s.get("date") == today]
    result["sections"]["struggles"] = today_struggles

    # 3. 今日洞察
    insights = hub.get_recent_insights(days=1)
    today_insights = [i for i in insights if i.get("date") == today]
    result["sections"]["insights"] = today_insights

    # 4. 情感记录
    moods = hub.get_recent_moods(days=1)
    today_moods = [m for m in moods if m.get("date") == today]
    if today_moods:
        result["sections"]["mood"] = today_moods[0]
    else:
        # 提示记录
        result["sections"]["mood_prompt"] = True

    # 事件记录
    hub.append_event(
        "life.reflection.evening",
        wins_count=len(today_wins),
        struggles_count=len(today_struggles),
        insights_count=len(today_insights)
    )

    return result


def format_reflection_text(result: dict, user_name: str = "主人") -> str:
    """
    将反思结果格式化为文本
    """
    now = datetime.now()
    date_str = now.strftime("%Y-%m-%d")

    sections = result.get("sections", {})

    lines = [
        f"晚上好，{user_name}。{date_str} 即将结束。",
        "",
    ]

    # 成就
    wins = sections.get("wins", [])
    if wins:
        lines.append("**今日成就**")
        for w in wins:
            lines.append(f"- 🎉 {w.get('description', '')}")
        lines.append("")

    # 困境
    struggles = sections.get("struggles", [])
    unresolved = [s for s in struggles if not s.get("resolved")]
    if unresolved:
        lines.append("**待解决**")
        for s in unresolved:
            lines.append(f"- 🤔 {s.get('description', '')}")
        lines.append("")

    # 洞察
    insights = sections.get("insights", [])
    if insights:
        lines.append("**今日洞察**")
        for i in insights:
            conf = i.get("confidence", 0)
            lines.append(f"- 💡 {i.get('observation', '')} (置信度: {conf:.0%})")
        lines.append("")

    # 情感提示
    if sections.get("mood_prompt"):
        lines.append("**情感记录**")
        lines.append("今天的心情如何？（1-10 分）")
        lines.append("")

    return "\n".join(lines)


# ========== Memory Consolidation ==========

def memory_consolidation(workspace_dir: str) -> dict:
    """
    23:00 - 记忆整合

    存储今日模式，更新习惯连续
    """
    hub = MemoryHub(workspace_dir)
    now = datetime.now()
    today = now.strftime("%Y-%m-%d")

    result = {
        "type": "memory_consolidation",
        "timestamp": now.isoformat(),
        "processed": []
    }

    # 1. 模式发现 - 基于今日数据
    moods = hub.get_recent_moods(days=1)
    today_moods = [m for m in moods if m.get("date") == today]

    if today_moods:
        avg_score = sum(m.get("score", 5) for m in today_moods) / len(today_moods)
        # 如果连续 3 天低于 6，生成洞察
        recent_moods = hub.get_recent_moods(days=3)
        if len(recent_moods) >= 3 and all(m.get("score", 5) < 6 for m in recent_moods):
            hub.record_insight(
                observation=f"连续 3 天情绪偏低（平均 {avg_score:.1f}/10），可能需要调整生活节奏",
                confidence=0.75,
                source="mood",
                pattern_based=True
            )
            result["processed"].append("mood_pattern_insight")

    # 2. 事件记录
    hub.append_event(
        "life.consolidation",
        date=today,
        patterns_count=len(result["processed"])
    )

    return result


# ========== Record Win ==========

def record_win(workspace_dir: str, description: str, category: str = "general") -> dict:
    """记录一个成就"""
    hub = MemoryHub(workspace_dir)
    result = hub.record_win(description=description, category=category)

    hub.append_event(
        "life.win.recorded",
        description=description,
        category=category
    )

    return result


# ========== Record Struggle ==========

def record_struggle(workspace_dir: str, description: str, resolved: bool = False) -> dict:
    """记录一个困境"""
    hub = MemoryHub(workspace_dir)
    result = hub.record_struggle(description=description, resolved=resolved)

    hub.append_event(
        "life.struggle.recorded",
        description=description,
        resolved=resolved
    )

    return result


# ========== CLI 入口 ==========

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: day-mode.py <workspace_dir> <action> [args...]")
        print("  Actions:")
        print("    briefing                    - Morning briefing")
        print("    checkin <level> [prod]      - Midday check-in (level: low|medium|high, prod: 1-10)")
        print("    reflection                  - Evening reflection")
        print("    consolidate                 - Memory consolidation")
        print("    win <description>           - Record a win")
        print("    struggle <description>      - Record a struggle")
        sys.exit(1)

    workspace_dir = sys.argv[1]
    action = sys.argv[2].lower()

    if action == "briefing":
        result = morning_briefing(workspace_dir)
        print(format_briefing_text(result))

    elif action == "checkin":
        level = sys.argv[3] if len(sys.argv) > 3 else None
        prod = int(sys.argv[4]) if len(sys.argv) > 4 else None
        note = sys.argv[5] if len(sys.argv) > 5 else ""
        result = midday_checkin(workspace_dir, level, prod, note)
        print(__import__('json').dumps(result, ensure_ascii=False, indent=2))

    elif action == "reflection":
        result = evening_reflection(workspace_dir)
        print(format_reflection_text(result))

    elif action == "consolidate":
        result = memory_consolidation(workspace_dir)
        print(__import__('json').dumps(result, ensure_ascii=False, indent=2))

    elif action == "win":
        if len(sys.argv) < 4:
            print("Error: win requires <description>")
            sys.exit(1)
        description = sys.argv[3]
        result = record_win(workspace_dir, description)
        print(f"Win recorded: {description}")

    elif action == "struggle":
        if len(sys.argv) < 4:
            print("Error: struggle requires <description>")
            sys.exit(1)
        description = sys.argv[3]
        result = record_struggle(workspace_dir, description)
        print(f"Struggle recorded: {description}")

    else:
        print(f"Unknown action: {action}")
        sys.exit(1)
